### ISSUE TITLE
Support for streams, buffers or file paths

### ISSUE THREAD

**Issue by jakubzitny** [TTLabs/EvaporateJS#199]
How difficult would it be to add support for buffers or file paths? So I would be able to upload not only instances of File from file form or drag and drop, but also from e.g. Electron's open file dialog (which returns strings)..

How much is EvaporateJS tied to files being Files? What about performance issues?

**Comment by bikeath1337** [comment]
The file passed to `Evaporate#add` is actually a JavaScript `File` which inherits from `Blob`. After a _cursory_ check, it looks like the only attribute of `file` that we use is `size` which is inherited from `Blob`. 

So you can try constructing a `Blob` from you get get from Electron and it will probably work. If not, you'll have clues about what might need to be done to support it.

**Comment by bikeath1337** [comment]
I think this can be closed as Evaporate uses the underlying Blob model for its multipart upload support.

**Comment by jakubzitny** [comment]
Well, I originaly wanted to have a possibility to use Evaporate under Node.js, where there is no `File` or `Blob` and I can work with files from string file paths with streams or non-Blob buffers.

To be more exact, we're using Electron to build a desktop app. In Electron you have a browser environment with access to Node and you can get files to upload from the "browser" part as `File`s, however, if I want to get files from OS file dialogs (e.g. from menu or right click menu) you never get a File, only a file path string. That's why I thought it might be cool for Evaporate to support that.

We've already worked around this and it's okay, but for the future it can still be a nice enhancement for Node devs.

**Comment by bikeath1337** [comment]
I just created a little test to demonstrate that Evaporate supports it now (unless I'm missing something). This sample creates a Blob from text but you can use any of the other methods to create a blob from other data.

```javascript
var jsonString = {hello: "world"},
  options = {type : "application/json"},
  blob = new Blob([JSON.stringify(jsonString, null, 2)], options),
  blobName = 'something random/my_blob';

  file_id = _e_.add({
    name: blobName,
    file: blob
  });
```

**Comment by jakubzitny** [comment]
@bikeath1337 but there is no `Blob` in Node. And even if there was, manually loading whole file into a string/array/buffer and then constructing a Blob from it is not really good for performance.

**Comment by bikeath1337** [comment]
It was to illustrate the point that what you need is a raw data buffer provided by some object. Evaporate uses FileReader to read that object in a very memory efficient way.

```javascript
getFilePart(me.file, part.start, part.end)
```

just returns the start and offset for the slice of bytes from the stream. I don't think this requires much change at all, to be honest, because we're simply slicing an array in all cases.

**Comment by jakubzitny** [comment]
The thing here is that slicing a `File` is (as you said) quite efficient a it'd be a pity to loose this when supporting something else. There is a way that could be similarly optimal and that could provide support for uploading from string file paths — streams.

Just to make sure you understand why is this interesting problem to solve — supporting something else than `File`s would allow Evaporate to expand to Node world and especially [Electron](http://electron.atom.io). Electron is web-based and has a `File` and it also has access to Node. Some Electron apis do not support Files though (e.g. [dialog](https://github.com/electron/electron/blob/master/docs/api/dialog.md) or [open-file event](https://github.com/electron/electron/blob/master/docs/api/app.md#event-open-file-macos)) and to upload them we have only their file path strings.

Now, when we have a string file path and Node's [fs](https://nodejs.org/api/fs.html) api, we can load the file into memory as a Buffer or as a Stream. Buffer is problematic when dealing with large files since we don't want to store all of their contents in memory. Streams could be better because you can only load a part of the file into memory and upload it one by one.

The replacement for `getFilePart` + some workarounds in other parts of the code could be enough, I'm going to look into this. The main thing could be something like this:

```js
const readStream = fs.createReadStream(me.file.path, { start: part.start, end: part.end });
const chunks = []
readStream.on('data', function(chunk) {
  chunks.push(chunk)
})
const blob = new Blob(Buffer.concat(chunks)) // -> the actual blob/slice/part to upload
```

**Comment by bikeath1337** [comment]
Yes, I poorly described that before but this is a summary of what I think the requirements should be, if we want this to work cooperatively and in the spirit of the original need:

- the solution should allow an object to be split into "parts"
- the solution should minimize it's memory footprint when processing those chunks

These two requirements are currently met with the File/FileReader implementations. The only known action that Evaporate takes which requires careful attention is when it calculates the MD5 checksum because it needs to have the whole part's content. There is a bit of code that specifically clears references to the chunk so that that part's memory can be cleaned up.

While I was developing the checksum support, I ran into several dead ends where either the CPU was maxed out calculating the checksums; or the asynchronous queue limits of the browser were hit or there was a memory leak caused by retaining unnecessary references to bits related to each part. Fortunately, it was possible to address all of them.

**Comment by bikeath1337** [comment]
This little enhancement really could be valuable and I'm looking forward to seeing what you create, @jakubzitny . Don't forget the specs!

**Comment by ikrtolica** [comment]
Can I second the request to try and add in streams support with node in mind?  The options for node S3 libraries are getting pretty bad as the official aws-sdk is a basket case and the other alternatives like knox are no longer under development.  I am willing to help in any way I can if needed.

**Comment by bikeath1337** [comment]
@ikrtolica thank you. Interesting to hear about other projects. I offered to help AWS with their SDK for JavaScript (to add the MD5 checksum support) and they accepted but I never followed up because I became engaged with Evaporate.

I was chatting with another developer user this week about how the MD5 checksums, the ability to upload small parts in parallel and the ability to resume after errors are incredibly powerful and would be really difficult to achieve without a lot of handshaking between client and server. That's why we now recommend using a smallish part size for the everyday user uploading from their average broadband connection.

But back to your offer to help. Yes, please. I may be naive but I believe that the secret to adding this feature is to develop against the Blob Interface because that is all that Evaporate requires to do the bulk of its work. I don't have any ideas yet if we're talking about a stream whose total length is unknown, though.

Any effort should not break backward compatibility, must include specs/tests and be based on v2.0.0.

**Comment by ikrtolica** [comment]
@bikeath1337 - glad to hear this is on your radar.  Node streams can be a little tricky to work with but I don't see why this could not be done.  Are you averse to using external libraries to as part of this project?  Seems like there are a couple projects out there that have attempted to provide a browser-like streaming interface to local files - for example there is an implementation of FileReader for node that seems to come close to what is needed here (https://www.npmjs.com/package/filereader).

**Comment by bikeath1337** [comment]
Thanks @ikrtolica . I would be open to a solution that did not impact the existing set of users who don't need this feature. I don't know if it's possible, but a solution similar to what we do for Promise support for v2.0 is require that developers include a Promise polyfill in their apps. Evaporate doesn't care who implemented the Promise. If we can do the same for FileReader, that could be straightforward to solve.

BTW, are these 'infinite' streams or finite streams or finite? In other words, Evaporate still needs to know when all parts are uploaded so it can construct and submit the Complete request.

**Comment by ikrtolica** [comment]
@bikeath1337 - maybe I spoke too soon - looks like the library I referenced is incomplete, so it will not work.  There are probably ways to make this work in node specifically, and probably the way to do this is with a polyfill so you can maintain features for existing users. However this is going to require some work to build out with the fs library in node.  It looks like the [knox-mpu](https://github.com/nathanoehlman/knox-mpu) project built multipart uploads using streams - the trick is to do your own chunking - they use lstat to find the size of a file for upload, then they stream it to disk in parts (if desired) and then stream each part to S3 in parallel.  Stream implementations do not usually give you a stream length, but we could keep track of parts in a similar way to the knox approach.

**Comment by bikeath1337** [comment]
_If_ there is a requirement for streams whose length is unknown when the upload starts, I think we need to start with enabling Evaporate to handle that. Such a change could easily be made to the code base without breaking existing functionality.

It would be great to have this question answered first.

**Comment by bikeath1337** [comment]
@ikrtolica @jakubzitny I have a working branch and POC that uses Node ReadableStream. After a lot of research, until Node decides to do the required marshaling, we have to do it on our own and this means a risk for increased memory consumption. Well, I try to manage that but if the partSize is huge and lots of parts are uploading there is going to be pressure on memory. Since I've been recommending smallish part sizes for better resumability, I'm fine with the risk.

There is plenty of documentation at the branch. I want to repeat that I did not hook into the native file system picker because it was too much work for me to learn electron. Instead, I used `fs.createReadableStream` using a file path. Thanks to @jakubzitny for the idea.

Thus, the electron example I provide, required me to upload files from my desktop. If someone wants to hook up a real file system, that would be nice but not required.

To get started with the readme, with installation instructions:  https://github.com/TTLabs/EvaporateJS/tree/electron

PR is based on a TK release of Evapoarate. I created the PR to illustrate what was required for stream support: https://github.com/bikeath1337/EvaporateJS/pull/4/files.

**Comment by bikeath1337** [comment]
Branch updated after release of 2.0.2

**Comment by bikeath1337** [comment]
Branch updated after release of 2.0.4 which reduces reported memory retention.

**Comment by bikeath1337** [comment]
First cut functionality included in v2.0.5
